va/va_prot.h

```diff
@@ -0,0 +1,456 @@
+ *     // not find desired counter mode
+ *     assert(0);
+ * }
+ * if ((attribs[4].value & VA_PC_SAMPLE_TYPE_SUBSAMPLE) == 0) {
+ *     // not find desired sample type
+ *     assert(0);
+ * }
+ * if ((attribs[5].value & VA_PC_USAGE_WIDEVINE) == 0) {
+ *     // not find desired usage
+ *     assert(0);
+ * }
+ * \endcode
+ *
+ * \section api_pc_setup Set up a protected content session
+ *
+ * The protected content session can be attached to VA decode/encode/vp context
+ * to do decryption/protection in the pipeline.
+ * Before creating a protected content session, it needs to create a config
+ * first via vaCreateConfig(). Then using this config id to create a protected
+ * content session via vaCreateProtectedSession().
+ *
+ * The general control flow is demonstrated by the following pseudo-code:
+ * \code
+ * // Create config
+ * VAConfigID config_id;
+ *
+ * attribs[0].value = VA_PC_CIPHER_AES;
+ * attribs[1].value = VA_PC_BLOCK_SIZE_128;
+ * attribs[2].value = VA_PC_CIPHER_MODE_CBC;
+ * attribs[3].value = VA_PC_SAMPLE_TYPE_SUBSAMPLE;
+ * attribs[4].value = VA_PC_USAGE_WIDEVINE;
+ * va_status = vaCreateConfig(va_dpy, VAProfileProtected,
+ *     VAEntrypointProtectedContent, attribs, 5, &config_id);
+ * CHECK_VASTATUS(va_status, "vaCreateConfig");
+ * \endcode
+ *
+ * Once the config is set up, we can create protected content session via
+ vaCreateProtectedSession().
+ * \code
+ * // Create a protected session
+ * VAProtectedSessionID crypto_session;
+ *
+ * va_status = vaCreateProtectedSession(va_dpy, config_id, &crypto_session);
+ * CHECK_VASTATUS(va_status, "vaCreateProtectedSession");
+ * \endcode
+ *
+ * \section api_pc_exec TEE communication via vaProtectedSessionExecute()
+ *
+ * Before start decryption/encryption, we may need to communicate with TEE to
+ * get encrypted decryption. We need to call vaProtectedSessionExecute() to
+ * get it. The following pseudo-code demonstrates getting session id via
+ * vaProtectedSessionExecute() as an example.
+ *
+ * \code
+ * uint32_t app_id = 0xFF;
+ * VABufferID buffer;
+ * VAProtectedSessionExecuteBuffer exec_buff = {0};
+ *
+ * exec_buff.function_id = GET_SESSION_ID
+ * exec_buff.input.data = nullptr;
+ * exec_buff.input.data_size = 0;
+ * exec_buff.output.data = &app_id;
+ * exec_buff.output.max_data_size = (uint32_t);
+ * va_status = vaCreateBuffer(
+ *                 va_dpy,
+ *                 crypto_session,
+ *                 (VABufferType) VAProtectedSessionExecuteBufferType,
+ *                 sizeof(exec_buff),
+ *                 1,
+ *                 &exec_buff,
+ *                 &buffer);
+ *
+ * va_status = vaProtectedSessionExecute(va_dpy, crypto_session, buffer);
+ *
+ * vaDestroyBuffer(va_dpy, buffer);
+ * \endcode
+ *
+ * \section api_pc_attach Attach/Detach protected content session to the VA
+ * context which want to enable/disable decryption/protection
+ *
+ * Protected content session is attached to VA decode/encode/vp context to
```

Comment:
It's still not clear why `vaProtectedSessionExecute` is needed in relevance to decoding process. You highlighted that you get some info via execute like app_id, but how do you use it? Specifically, are we supposed to attach any of the information you get via vaProtectedSessionExecute as any buffer via `vaRenderPicture`? If yes - how (do we have any particular buffer type for that)? Is that why VAEncryptionParameterBufferType is needed for?

Reply:
I still think this question is not answered in the header text. Basically: does any of the assets you get via `vaProtectedSessionExecute` end up filled in `VAEncryptionParameterBufferType`? which and how?

Reply:
Yes. 
Using vaProtectedSessionExecute, Application obtains IHV specific TEE binary blob for protected session initialization. This binary blob is subsequently used by Application(using VAEncryptionParameterBufferType) to provide necessary cryptographic/security information to GPU to prime HWDRM session that adhere to licensing requirements.
Additionally, vaProtectedSessionExecute is further used by Application to obtain HWDRM specific TEE binary blob for protected session playback. This binary blob is subsequently used by Application(using VAEncryptionParameterBufferType) to provide necessary cryptographic/security information to GPU to enforce spatial(which machine license is bound)/temporal(how long license is valid) requirements on HWDRM session in accordance with licensing requirements.